Feature/add work package attachments (WP 29203)

virgo47 opened the pull request:
I'd like to contribute the implementation of [WP 29203](https://community.openproject.org/projects/openproject/work_packages/29203/activity), feature "Allow attachment upload on read-only work packages". The implementation is based on a new permission `:add_work_package_attachments` which works on its own or can be combined with adding notes (comments) for drag'n'drop functionality - this works as expected automatically when the attachment links are provided in the API responses.

I managed to extend one test only `spec/features/work_packages/attachments/attachment_upload_spec.rb` - and was lucky enough to conflict with incoming changes in the process. ;-)

Guys, please, give me some hints what other tests I should extend or what new tests (ideally similar to an existing ones) I should add, because I'm quite struggling to orient myself in the tests.

As for manual validation, the process is easy:
* If the member has Edit WP permission, everything works as before.
* If the member does not have Edit WP nor the new permission, everything works as before (attachments can't be added to an existing package).
* If the member has the new permission and no Edit WP (e.g. some "Attachment adder" role) then the user can add attachments to an existing WP:
  * They can d'n'd to Files area.
  * They can use the Attach button there to add it via OS open file dialog.
  * IF the role has also permission to add comments ("notes"), then they can also d'n'd into the new comment.

The new permission does not provide delete attachment capability.

virgo47 committed:
initial implementation of add_work_package_attachments permission

This allows adding attachments even without Edit WP permission.

virgo47 committed:
Merge branch 'opf:dev' into feature/add_work_package_attachments

virgo47 committed:
Merge branch 'opf:dev' into feature/add_work_package_attachments

virgo47 committed:
Merge branch 'opf:dev' into feature/add_work_package_attachments

virgo47 committed:
wp/attachment_upload_spec: added add_work_package_attachments tests

virgo47 commented:
I'm on those `./spec/requests/api/v3/attachments/work_package_spec.rb` failures.

virgo47 committed:
Merge branch 'opf:dev' into feature/add_work_package_attachments

virgo47 committed:
acts_as_attachable#attachments_addable? now supports permission arrays

- this is necessary to support :add_work_package_attachments permission
- permission arrays were already supported by allowed_to_on_attachment?
- this fixes attachments/work_package_spec.rb

oliverguenther commented:
Hi @virgo47 , thanks for the contribution!

cbliard commented on config/initializers/permissions.rb:
```diff
@@ -309,6 +309,12 @@
+      wpt.permission :add_work_package_attachments,
+                     {},
+                     permissible_on: %i[work_package project],
+                     dependencies: :view_work_packages
```
Can you please move this permission definition after the `:edit_own_work_package_notes` so it is grouped with other similar permissions.

cbliard commented on config/locales/en.yml:
```diff
@@ -2475,6 +2475,8 @@ en:
   permission_add_messages: "Post messages"
   permission_add_project: "Create projects"
+  permission_add_work_package_attachments: "Add attachments"
+  permission_add_work_package_attachments_explanation: "Allows adding attachments without Edit work package permission"
```
There is a missing s at the end of "Edit work package**s**".

cbliard reviewed (changes requested):
Thank you for your contribution!

The functional tests you added cover the parts you described. If you want to also cover the "The new permission does not provide delete attachment capability." part, you could add a test `spec/requests/api/v3/work_packages/delete_resource_spec.rb` to ensure having `:add_work_package_attachments` is not enough to delete a work package, but I'm not sure if that's really needed.

I also have some minor enhancements requests about your code. I'll merge the PR once you fix them.

virgo47 committed:
Merge branch 'dev' into feature/add_work_package_attachments

virgo47 committed:
review fixes and test enhancements for PR #13861

- Add_work_package_attachments permission moved to other
similar permissions.
- Localization typo fix
- wp/{delete,update}_resource_specs also try new permission

virgo47 commented:
Thanks for the review! I've made the suggested changes and also added the permission to the negative scenario in `.../work_packages/delete_resource_spec.rb`, and in the same manner for `update_resource_spec` as well. I hope you don't mind using two permissions in the test at once, but I believe the test idea stayed the same. Repeating it again in a different context didn't make sense to me - but let me know what you think. ;-)
I also merged recent dev into the branch.

cbliard commented:
> I hope you don't mind using two permissions in the test at once, but I believe the test idea stayed the same. Repeating it again in a different context didn't make sense to me

You're right.

Thanks a lot for implementing this. You pull request was well tested and was a pleasure to review :heart:
This feature will be released in version 13.1.
